File: 509.fibonacci-number.kt
// @leet start
class Solution {
    fun fib(n: Int): Int {
      fun fib(n: Int, memo: IntArray): Int {
        if (n < 2) return n
        if (memo[n] != 0) return memo[n]
        memo[n] = fib(n - 1, memo) + fib(n - 2, memo)
        return memo[n]
      }
      return fib(n, IntArray(n + 1))
    }
}
// @leet end

File: 509.fibonacci-number.go
// @leet start
func fib(n int) int {
  if n < 2 { return n }
  m := make([]int, 2)
  m[0], m[1] = 0, 1
  for i := 2; i <= n; i++ {
    r := m[0] + m[1]
    m[0], m[1] = m[1], r
  }
  return m[1]
}
// @leet end
